hal/src/thumbv6m/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1353 @@
+
+        Self {
+            sercom,
+            pads,
+            chsize: PhantomData,
+            freq: freq.into(),
+        }
+    }
+}
+
+impl<P, C> Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    /// Reset the SERCOM peripheral
+    #[inline]
+    fn swrst(sercom: &P::Sercom) {
+        sercom.usart().ctrla.write(|w| w.swrst().set_bit());
+        while sercom.usart().syncbusy.read().swrst().bit_is_set() {}
+    }
+
+    /// Change the [`Config`] [`CharSize`]
+    #[inline]
+    fn change<C2>(self) -> Config<P, C2>
+    where
+        C2: CharSize,
+    {
+        Config {
+            sercom: self.sercom,
+            pads: self.pads,
+            chsize: PhantomData,
+            freq: self.freq,
+        }
+    }
+
+    /// Trigger the [`Sercom`]'s SWRST and return a [`Config`] in the
+    /// default configuration.
+    #[inline]
+    pub fn reset(self) -> Config<P> {
+        Config::create(self.sercom, self.pads, self.freq)
+    }
+
+    /// Consume the [`Config`], reset the peripheral, and return the [`Sercom`]
+    /// and [`Pads`]
+    #[inline]
+    pub fn free(self) -> (P::Sercom, P) {
+        Self::swrst(&self.sercom);
+        (self.sercom, self.pads)
+    }
+
+    /// Change the [`CharSize`]
+    #[inline]
+    pub fn char_size<C2: CharSize>(self) -> Config<P, C2> {
+        C2::configure(&self.sercom);
+        self.change()
+    }
+
+    /// Change the bit order of transmission (MSB/LSB first)
+    #[inline]
+    pub fn msb_first(mut self, msb_first: bool) -> Self {
+        self._msb_first(msb_first);
+        self
+    }
+
+    /// Change the parity setting
+    #[inline]
+    pub fn parity(mut self, parity: Option<Parity>) -> Self {
+        self._parity(parity);
+        self
+    }
+
+    /// Change the stop bit setting
+    #[inline]
+    pub fn stop_bit(mut self, stop_bit: StopBits) -> Self {
+        self._stop_bit(stop_bit);
+        self
+    }
+
+    /// Enable or disable the start of frame detector.
+    ///
+    /// When set, the UART will generate interrupts for
+    /// RXC and/or RXS if these interrupt flags have been enabled.
+    #[inline]
+    pub fn start_of_frame_detection(mut self, enabled: bool) -> Self {
+        self._start_of_frame_detection(enabled);
+        self
+    }
+
+    /// Enable or disable the collision detector.
+    ///
+    /// When set, the UART will detect collisions and update the
+    /// corresponding flag in the STATUS register.
+    #[inline]
+    pub fn collision_detection(mut self, enabled: bool) -> Self {
+        self._collision_detection(enabled);
+        self
+    }
+
+    /// Set the baud rate
+    ///
+    /// This function will calculate the best BAUD register setting based on the
+    /// stored GCLK frequency and desired baud rate. The maximum baud rate is
+    /// GCLK frequency/oversampling. Values outside this range will saturate at
+    /// the maximum supported baud rate.
+    ///
+    /// Note that 3x oversampling is not supported.
```

Comment:
I have not read about UART baud rates *at all*, but I'm curious why this isn't supported.

Reply:
I made the decision not to support it. The only reason is that 3x oversampling only supports arithmetic baud mode, and not fractional mode. Supporting it would've added complexity to the baud mode enum system. Maybe this isn't a good reason though.

Reply:
I think it's fine. I chose to ignore 9-bit character sizes for the `thumbv7` `spi` module. Same reasoning, added complexity for little benefit.